src/planner/binder/statement/bind_insert.cpp

```diff
@@ -115,18 +119,76 @@ BoundStatement Binder::Bind(InsertStatement &stmt) {
 		}
 	}
 
-	// insert from select statement
 	// parse select statement and add to logical plan
 	auto root_select = Bind(*stmt.select_statement);
 	CheckInsertColumnCountMismatch(expected_columns, root_select.types.size(), !stmt.columns.empty(),
 	                               table->name.c_str());
 
 	auto root = CastLogicalOperatorToTypes(root_select.types, insert->expected_types, move(root_select.plan));
+
 	insert->AddChild(move(root));
 
-	result.plan = move(insert);
-	this->allow_stream_result = false;
-	return result;
+	if (!stmt.returning_list.empty()) {
+		insert->return_chunk = true;
+		// clear result type and names because returning list will update them
+		result.types.clear();
+		result.names.clear();
+
+		auto insert_table_index = GenerateTableIndex();
+		insert->table_index = insert_table_index;
+
+		auto binder = Binder::CreateBinder(context);
+		auto types = table->GetTypes();
+		auto names = vector<std::string>();
+		for (auto &col : table->columns) {
+			names.push_back(col.name);
+		}
+
+		binder->bind_context.AddGenericBinding(insert->table_index, stmt.table, names, types);
+		InsertBinder insert_binder(*binder, context);
+
+		unique_ptr<LogicalOperator> insert_as_logicaloperator = move(insert);
+
+		auto projection_expressions = vector<unique_ptr<Expression>>();
+		LogicalType result_type;
+		for (auto &returning_expr : stmt.returning_list) {
+			auto expr_type = returning_expr->GetExpressionType();
+			if (expr_type == ExpressionType::STAR) {
+				auto generated_star_list = vector<unique_ptr<ParsedExpression>>();
+				binder->bind_context.GenerateAllColumnExpressions((StarExpression &)*returning_expr,
+				                                                  generated_star_list);
+
+				for (auto &star_column : generated_star_list) {
+					auto star_expr = insert_binder.Bind(star_column, &result_type);
+					result.types.push_back(result_type);
+					result.names.push_back(star_expr->ToString());
+					projection_expressions.push_back(move(star_expr));
+				}
+			} else {
+				auto expr = insert_binder.Bind(returning_expr, &result_type);
+				result.names.push_back(expr->ToString());
+				result.types.push_back(result_type);
+				if (expr_type != ExpressionType::BOUND_COLUMN_REF) {
+					PlanSubqueries(&expr, &insert_as_logicaloperator);
+				}
+				projection_expressions.push_back(move(expr));
+			}
+		}
+
+		auto projection = make_unique<LogicalProjection>(insert_table_index, move(projection_expressions));
```

Comment:
Could we add some more tests:

* RETURNING statement on a table with different types (e.g. `a varchar, b int`) - and using the types in different orders
* RETURNING combinations of columns (`a+b`)
* RETURNING constant values (`'hello'`)
* RETURNING casts to different types (`a::VARCHAR`)
* RETURNING lists (`[a, b, c]`)
* RETURNING structs (`{'a': a, 'b: b, 'c': c}`)
* RETURNING with a default value from a sequence, e.g.:
* RETURNING with subqueries
```sql
statement ok
CREATE SEQUENCE seq;

statement ok
CREATE TABLE test (a INTEGER DEFAULT nextval('seq'), b INTEGER);
```

* More error conditions (RETURNING with unknown columns, RETURNING with window functions)